### PR TITLE
Part 1: Refactor upload

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
This does not include any changes to the iOS code related to uploading. I will need to riddle out how to set up mocks for the iOS calls (and/or figure out how to rewrite it all in Kotlin using expect/actually) but I wanted to get some eyes on this since I pulled out the easily shared code from the Android-only library.

### CODE REVIEW

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt [hunk 1]
```diff
@@ -23,14 +24,22 @@ data class UploadFile (
         return "uploadSession--$filePath"
     }
 
+    internal fun getSecondaryId(): String {
+        return metadata?.instanceGuid ?: filename()
+    }
+
+    internal fun filename(): String {
+        return filePath.subSequence(filePath.lastIndexOf("/")+1, filePath.length).toString()
```
Comment (Erin-Mounts) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt:32]: Does Kotlin not have built in methods for parsing out the components of file paths/URLs like iOS does?

**Review by Erin-Mounts — APPROVED** [review]
One question, otherwise looks good to me

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt (hunk 1 above)
Comment (syoung-smallwisdom) [reply bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt:32]: This is copy/paste - I'm guessing it doesn't though yeah, it's a bit icky.

Comment (nategbrown9) [reply bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UploadFile.kt:32]: The java File API would do this, but it is only available for Android/JVM. We now might be able to use something from Okio to do this now.

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+import org.sagebionetworks.bridge.kmm.shared.cache.loadResource
+import org.sagebionetworks.bridge.kmm.shared.models.UploadFile
+import org.sagebionetworks.bridge.kmm.shared.models.UploadRequest
+import org.sagebionetworks.bridge.kmm.shared.models.UploadSession
+
+internal open class UploadRepo(
+    httpClient: HttpClient,
+    databaseHelper: ResourceDatabaseHelper,
+    backgroundScope: CoroutineScope,
+) : AbstractResourceRepo(databaseHelper, backgroundScope) {
```
Comment (nategbrown9) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/UploadRepo.kt:28]: I don't think we gain anything by extending AbstractResourceRepo here.

**Review by nategbrown9 — APPROVED** [review]
Looks good.

**Pull request merged**